capstone/src/main/java/com/step902020/capstone/Recommender.java

```diff
@@ -0,0 +1,70 @@
+
+package com.step902020.capstone;
+
+import com.step902020.capstone.security.CurrentUser;
+import java.io.IOException;
+import java.util.*;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.util.Pair;
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.RequestParam;
+import org.springframework.web.bind.annotation.RequestHeader;
+import org.springframework.web.bind.annotation.RestController;
+import org.springframework.web.servlet.view.RedirectView;
+import com.google.gson.Gson;
+import java.io.IOException;
+import java.util.function.Function;
+
+public class Recommender {
+
+  public static <E, U> List<E> recommend (U targetUser, List<U> users, Function<U, List<E>> getItemList) {
+    users.remove(targetUser);
+    Map<U, Double> userToScore= new HashMap<U, Double>();
+    Set<E> targetUserEvents = new HashSet(getItemList.apply(targetUser));
+    for (U user: users) {
+      int dist = 0;
+      Set<E> userEvents = new HashSet(getItemList.apply(user));
+      for (E e : targetUserEvents) {
+        if (!userEvents.contains(e)) {
+          dist++;
+        }
+      }
+      for (E e : userEvents) {
+        if (!targetUserEvents.contains(e)) {
+          dist++;
+        }
+      }
+      double scaledDist = 1/(1+Math.sqrt(dist));
```

Comment:
Giving it a bit more thought, I think your distance calculation is actually correct and equivalent to the euclidian distance. However, I'm still not sure about the whole scaling part.

Reply:
I agree the scaling part would not actually make a difference. However, I do have some concerns about the euclidean distance. Currently what I am doing is considering only events that either one or the other user have added. This means that for each user I do not consider their distance with respect to the entire event space. 
Example: 
- the list of all events: A, B, C, D
- user 1 saves A, B
- user 2 saves A, C
in this case the dimension would be 3 instead of 4 because in my distance calculation I only consider A, B, C
Another example:
- user 1 saves A,
- user 2 saves A, B
in this case the dimension I'm considering would be 2 because I only consider distance WRT to A, B

Is this a valid way of doing it? The differing dimensions are concerning me

Reply:
Well I think the scaling actually does make a difference because it inverts the values. If you start out with a distance between u1 & u2 as 100 and u1 & u3 as 1000, the scaling makes the distance between u1 & u2 as 1/sqrt(100)=1/10=**.1** and between u1 & u3 as 1/sqrt(1000)=1/100=**.01**. So, the smaller distance became the bigger distance.

In your first example:
u1 = [1, 1, 0, 0]
u2 = [1, 0, 1, 0]
d(u1, u2) = sqrt((1-1)^2 + (1-0)^2 + (0-1)^2 + (0-0)^2) = sqrt(0 + 1 + 1 + 0) = sqrt(2)

Second example:
u1 = [1, 1, 0, 0]
u2 = [1, 0, 0, 0]
d(u1, u2) = sqrt((1-1)^2 + (1-0)^2 + (0-0)^2 + (0-0)^2) = sqrt(0 + 1 + 0 + 0) = sqrt(1)

As you can, see you don't need to worry about events that exist but neither u1 or u2 save because they become 0s in the formula. The only events you need to add up are the ones where u1 saved and u2 didn't and vice versa, which is exactly what your algorithm is doing as far as I understand. BTW, the final sqrt doesn't really matter because it doesn't change the relative ranking.

Reply:
Ohh yes I see. In this case I won't use the scaling and will just sort in ascending order.